lua/telescope/utils.lua

```diff
@@ -284,4 +284,244 @@ function utils.strcharpart(str, nchar, charlen)
+
+    if exit_visual_mode then
+      vim.cmd [[normal :esc<CR>]]
+    end
+
+    -- Resolution via wincol (see :help wincol) enables reconstruction of blockwise selection
+    local wincol_start, wincol_end, wincol_left, wincol_right, c_flip, l_flip
+    if mode == '' then
+        wincol_start = utils.get_wincol({l_start, c_start - 1}, {l_start, c_start - 1})
+        wincol_end = utils.get_wincol({l_end, c_end - 1}, {l_start, c_start -1 })
+        -- backwards visual selection
+        -- visual-block mode switch only after getting wincol column edges!
+        if l_start > l_end then
+            l_end, l_start = l_start, l_end
+            l_flip = true
+        end
+        if c_start > c_end then
+            c_end, c_start = c_start, c_end
+            c_flip = true
+        end
+        -- Blockwise selection requires the very most edge wincol
+        -- Consider a multi-width char: 全, wincol solely returns the first visual column though char occupies 2 cols
+        -- edge_wincol returns the left or right border wincol instead depending on 'maximum' param
+            -- get edge pos for wincol start (most left) and end (most right) by line from original positions
+            -- flipping is required to know what direction to resolve to
+            wincol_start = utils.edge_wincol({l_flip and l_end or l_start,          -- get l_end if flipped else original l_start
+                                              c_flip and c_end - 1 or c_start - 1},  -- get c_end if flipped else original c_start
+                                              wincol_start,                            -- resolve for wincol_start
+                                              c_flip)                                -- flipped (true) or not (false), resolve to true=right, left=false
+            wincol_end = utils.edge_wincol({l_flip and l_start or l_end,            -- resolve conversely in opposite direction as above
+                                            c_flip and c_start - 1 or c_end - 1},
+                                            wincol_end,
+                                            not c_flip)
+            -- properly set left and right edge wincol
+            wincol_left = math.min(wincol_start, wincol_end)
+            wincol_right = math.max(wincol_start, wincol_end)
+        return l_start, c_start, l_end, c_end, wincol_left, wincol_right
+    else
+      -- char- or linewise selection
+      if l_start > l_end or (l_start == l_end and c_end < c_start) then
+        return l_end, c_end, l_start, c_start
+      end
+      return l_start, c_start, l_end, c_end
+    end
+end
+
+-- Retrieve wincol from pos
+-- wincol (see :help wincol) denotes the visual column of the current cursor position
+-- Since wincol can only be retrieved for the current position,
+-- the cursor has to be set prior
+-- @param pos table position to retrieve wincol for
+-- @param start_pos table optional position to jump back to
```

Comment:
Since `get_wincol` does the jump, shouldn't it also be responsible for restoring cursor position?

Reply:
Not sure I follow. If `start_pos` is set the cursor jumps back to that position. In the actual function `get_visual_selection`, I also jump back right before returning the visual selection as otherwise I'd have to keep track of the starting position everywhere (though I'm sure it currently jumps more often than needed :) - will check what can be removed.

Reply:
If I understand correctly (and I might not!), what currently happens is:

1. the calling function (`get_visual_selection`) stores the current cursor position
2. it then calls `get_wincol` with that position as argument
3. `get_wincol` jumps to the location of the mark and then jumps back to the passed position

I'd suggest to have `get_wincol` _always_ jump back (saving the current position before jumping), even if that amounts to duplicating saving the current position. Otherwise it seems a bit of a misnomer: I'd expect a function called `get_wincol` to be side-effect free so it's guaranteed to return the position without changing state (like changing cursor position).

Reply:
Thanks for elaborating, makes perfect sense! Your understanding is correct, `vim.fn.wincol` only returns the visual column from the current cursor position. Other than that, speed is a non-issue (I guess thanks to luajit) as e.g. 500 loc are  processed instantaneously.